### PR TITLE
Mempool concurrency: transaction validation

### PR DESCRIPTION
**Pull request by kirushanths**
@mslipper

### CODE REVIEW

File: pkg/service/mempool.go [hunk 1]
```diff
@@ -50,36 +51,58 @@ type flushSpendReq struct {
...
 }
 
+func (m *Mempool) validateTransactionRequest(req *txRequest) {
+        tx := req.tx
+        var err error
+        if tx.Body.IsDeposit() {
+                err = m.VerifyDepositTransaction(&tx)
+        } else {
+                // 300-500us
+                err = m.VerifySpendTransaction(&tx)
```
Comment (kirushanths) [inline pkg/service/mempool.go]: Once you move out the `mempool.poolSpends` checks out of this method and it just becomes `validation.ValidateSpendTransaction` (which only checks the previous block in storage and does math only, it basically can be parallelized).

I know you said the `storage.isDoubleSpent` can **not** be parallelized but exactly why? it's checking the previous _persisted_ block only, which won't change until mempool gets flushed. And Mempool basically contains transactions for the next block and having the `ensurePoolSpends` separated on a single thread (under the new `case req := <-m.validatedTxReqs:`) ensures no double-spend happens in mempool.

In summary: `storage` doubleSpend check can be parallelized cause it's past history and static but maintaining mempool double-spend integrity on a single thread keeps the correctness.

Thoughts?

File: pkg/eth/signatures.go [hunk 2]
```diff
@@ -21,22 +21,27 @@ func Sign(privKey *ecdsa.PrivateKey, hash util.Hash) (chain.Signature, error) {
...
 
+        // 200 ns
 	sigCopy := make([]byte, len(signature))
 	copy(sigCopy, signature)
 	if len(sigCopy) == 65 && sigCopy[64] > 26 {
 		sigCopy[64] -= 27
 	}
 
+        // 220 Us
 	pubKey, err := crypto.SigToPub(ethHash, sigCopy)
```
Comment (kirushanths) [inline pkg/eth/signatures.go]: Basically 70% of the work happens right here on this line.

File: pkg/service/mempool.go [hunk 3]
```diff
@@ -50,36 +51,58 @@ type flushSpendReq struct {
...
-				tx := req.tx
-				var err error
-				if tx.Body.IsDeposit() {
-					err = m.VerifyDepositTransaction(&tx)
-				} else {
-					err = m.VerifySpendTransaction(&tx)
-				}
-				if err != nil {
+                                go m.validateTransactionRequest(req)
+                        case req := <-m.validatedTxReqs:
```
Comment (kirushanths) [inline pkg/service/mempool.go]: We validate the incoming transaction request first against previous block history for double-spends and once that's cleared we do our mem-pool integrity checks synchronously below.

**Comment by kirushanths** [comment]
Also if you can configure CircleCI to run all commits (including forked PRs) would be great.

File: pkg/validation/validation.go [hunk 4]
```diff
@@ -14,31 +14,38 @@ import (
...
+        // 200 Us
 	prevTx0Output := prevTx0.Body.OutputAt(tx.Body.Input0.OutputIndex)
 	err = eth.ValidateSignature(sigHash0, tx.Sigs[0][:], prevTx0Output.Owner)
 	if err != nil {
 		return NewErrInvalidSignature(0)
 	}
 
+        // 450 ns
 	totalInput := big.NewInt(0)
 	totalInput = totalInput.Add(totalInput, prevTx0Output.Amount)
```
Comment (kirushanths) [inline pkg/validation/validation.go]: Another thing: `if !tx.Body.Input1.IsZero()` never gets tripped which will essentially could double the time because `eth.valdiateSignature` gets called twice.

File: pkg/eth/signatures.go (hunk 2 above)
Comment (kirushanths) [reply pkg/eth/signatures.go]: If you do the rough math here: If this takes say on average 200 us. That means about 5 transactions processed a millisecond. And in a 100ms block with 16-cores (my setup0 parallelizing the processing you get: 5*100*16 = 8000 TPS. I'm basically this figure plus/minus overhead.

File: pkg/service/mempool.go (hunk 1 above)
Comment (kirushanths) [reply pkg/service/mempool.go]: Second thoughts on this: realized that if a block is about to be written with transaction X (via flush), but a double-spend transaction X' is in a go-routine just past this line, it's possible for a double-spend to sneak in.

Basically mempool validation can't occur while a block is about to be written.

So you're right - I will need to re-think a little bit more

File: pkg/validation/validation.go (hunk 4 above)
Comment (mslipper) [reply pkg/validation/validation.go]: What do you mean never gets tripped? You mean if `Input1` is defined?

File: pkg/service/mempool.go (hunk 1 above)
Comment (mslipper) [reply pkg/service/mempool.go]: Correct, mempool validation cannot occur while a block is about to be/is being written. That's handled by the `flushReq` case in the mempool loop - all transaction processing halts until the block is successfully written to disk.

File: pkg/service/mempool.go (hunk 3 above)
Comment (mslipper) [reply pkg/service/mempool.go]: I like this approach.

File: pkg/service/mempool.go (hunk 1 above)
Comment (mslipper) [reply pkg/service/mempool.go]: When I said the double spend check can't be parallelized I was referring to how it requires that blocks be written to disk before new transactions can be processed by the mempool.

**Comment by mslipper** [comment]
Going to try reopening to kick CI.

**Comment by mslipper** [comment]
There it goes.

File: pkg/validation/validation.go (hunk 4 above)
Comment (kirushanths) [reply pkg/validation/validation.go]: Yeah if `input1` is never defined. which is the case in our benchmark setup

File: pkg/service/mempool.go (hunk 1 above)
Comment (kirushanths) [reply pkg/service/mempool.go]: Yeah but I think there is a race-condition introduced here. As described in my last comment: https://github.com/kyokan/plasma/pull/103/files#r259601343.

The transaction validation can occur simultaneously as a package block write. Basically we should wrap the go-routine in a lock.
